crates/aptos-telemetry/src/sender.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::metrics;
+use anyhow::{anyhow, Error};
+use aptos_config::config::NodeConfig;
+use aptos_crypto::{
+    noise::{self, NoiseConfig},
+    x25519,
+};
+use aptos_infallible::{Mutex, RwLock};
+use aptos_logger::{debug, error};
+use aptos_telemetry_service::types::{
+    auth::{AuthRequest, AuthResponse},
+    telemetry::TelemetryDump,
+};
+use aptos_types::{chain_id::ChainId, PeerId};
+use reqwest::{StatusCode, Url};
+use std::sync::Arc;
+
+// const METRICS_ENDPOINT: &str = "https://aptos.dev/telemetry/";
+// const AUTH_URL: &str = "https://aptos.dev/telemetry/auth";
+// const METRICS_URL: &str = "https://aptos.dev/telemetry/telemetry";
+
+const METRICS_ENDPOINT: &str = "http://localhost:8081/";
+const AUTH_URL: &str = "http://localhost:8081/auth";
+const METRICS_URL: &str = "http://localhost:8081/telemetry";
+const MAX_RETRIES: i32 = 5;
+
+struct AuthContext {
+    noise_config: Option<NoiseConfig>,
+    token: RwLock<Option<String>>,
+    server_public_key: Mutex<Option<x25519::PublicKey>>,
+}
+
+impl AuthContext {
+    fn new(node_config: &NodeConfig) -> Self {
+        Self {
+            noise_config: node_config.identity_key().map(NoiseConfig::new),
+            token: RwLock::new(None),
+            server_public_key: Mutex::new(None),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct TelemetrySender {
+    chain_id: ChainId,
+    peer_id: PeerId,
+    client: reqwest::Client,
+    auth_context: Arc<AuthContext>,
+}
+
+impl TelemetrySender {
+    pub fn new(chain_id: ChainId, node_config: &NodeConfig) -> Self {
+        Self {
+            chain_id,
+            peer_id: node_config.peer_id().unwrap_or(PeerId::ZERO),
+            client: reqwest::Client::new(),
+            auth_context: Arc::new(AuthContext::new(node_config)),
+        }
+    }
+
+    pub async fn send_metrics(&self, event_name: String, telemetry_dump: TelemetryDump) {
+        for i in 1..MAX_RETRIES {
```

Comment:
Better to use a retry policy here with some exponential backoff - https://github.com/aptos-labs/aptos-core/blob/e17fa8823787cd81fba22aebaf30f71ff4466411/crates/transaction-emitter-lib/src/emitter/mod.rs#L60

Reply:
Yes, using tokio-retry because of async'ness.